task/thermal/src/control.rs

```diff
@@ -116,84 +142,216 @@ impl InputChannel {
 /// This object uses slices of sensors and fans, which must be owned
 /// elsewhere; the standard pattern is to create static arrays in a
 /// `struct Bsp` which is conditionally included based on board name.
-///
-/// ```
-///         |
-///   HOT   |   Fan speed increases quickly
-///         |
-///         0------ Component margin + target_margin --
-///         |
-///         |   Fan speed increases slowly
-///         |
-///         ------- slow_band -------------------------
-///         |
-///         |   Fan speed remains constant
-///         |
-///         ------- dead_band -------------------------
-///         |
-///   COLD  |   Fan speed decreases slowly
-///         |
-/// ```
-///
-pub(crate) struct ThermalControl<'a, B> {
+pub(crate) struct ThermalControl<'a> {
     /// Reference to board-specific parameters
-    bsp: &'a B,
+    bsp: &'a Bsp,
 
     /// Task to which we should post sensor data updates
     sensor_api: SensorApi,
 
-    /// Dead band between increasing and decreasing fan speed
-    dead_band: Celsius,
-
-    /// Band where fan speed increases slowly
-    slow_band: Celsius,
-
     /// Target temperature margin. This must be >= 0; as it increases, parts
-    /// are kept cooler than their max temperature ratings.
+    /// are kept cooler than their target temperature value.
     target_margin: Celsius,
 
-    /// Commanded PWM value (0-100) for every output channel
-    target_pwm: u8,
+    /// Controller state
+    state: ThermalControlState,
+
+    /// How long to wait in the `Overheated` state before powering down
+    overheat_timeout_ms: u64,
 
+    /// Once we're in `Overheated`, how much does the temperature have to drop
+    /// by before we return to `Normal`
+    overheat_hysteresis: Celsius,
+
+    /// Most recent power mode mask
+    power_mode: u32,
+
+    /// Number of sensor reads which failed
     read_failed_count: u32,
```

Comment:
I haven't read where this is used yet. Given that I'm speculating on the implementation, maybe a sensor is marked as being mandatory for particular power states. A count here makes me think about there being a threshold, but this could just be a statistic.

Reply:
I'm going to remove `read_failed_count` and `post_failed_count`, since they aren't actually _read_ anywhere.

For now, we'll log read and post errors to the ringbuf; eventually, we'll need to come up with a way to ship that data to MGS / Nexus.